### ISSUE TITLE
BigQuery `jobs.insert` needs both `upload` (for `configuration.load`) and `doit` (for `configuration.query`, `.extract` and `.copy`)

### ISSUE THREAD

**Issue by emk** [Byron/google-apis-rs#191]
Hello! I've been experimenting with BigQuery and Rust and I've been I've been really impressed with the `google-bigquery2` crate. I managed to get OAuth2 and some simple queries working, and everything worked beautifully.

However, I ran into an interesting complication. The `hub.jobs().query(...)` is meant for small queries that output a few hundred or a few thousand rows of JSON. Larger queries need to be run using `jobs().insert(...)` with a `configuration.query` structure, and the results need to be exported to Google Cloud Storage using the same API with a `configuration.extract` structure.

But this means that `JobInsertCall` actually needs _both_ an `upload` method, and a `doit` method! Specifically, if you look at [the documentation here](https://cloud.google.com/bigquery/docs/reference/rest/v2/jobs/insert), the breakdown appears to be:

- `configuration.load`: Needs `upload`.
- `configuration.query`: Needs a regular `doit`.
- `configuration.extract`: Needs a regular `doit`.
- `configuration.copy`: Needs a regular `doit`.

But `JobInsertCall` has a private `doit` method that only provides an implementation for `upload` using `POST https://www.googleapis.com/upload/bigquery/v2/projects/projectId/jobs`, and which has no access to regular `doit` using `POST https://www.googleapis.com/bigquery/v2/projects/projectId/jobs`.

The key code appears to be in [mbuild.mako](https://github.com/Byron/google-apis-rs/blob/7ef8049d2907b54f71566d5e8fb69538deed3a0c/src/mako/api/lib/mbuild.mako#L396-L891), which is set up to generate _either_ a public `doit` function, or both a public `upload` function and a private `doit` function that can only be used to help implement `upload`.

The error that BigQuery gives me is:

```
Error: Bad Request (400): Uploads are only allowed with load jobs
    global: Uploads are only allowed with load jobs, invalid
```

I would be happy to submit a PR, but this involves a refactoring of `mbuild.mako`. Is this something that you would prefer to do yourself? Or would you be willing to explain how you'd like the code to be refactored?

I'm going to go ahead and modify my local copy to get rid of `upload_action` on this endpoint, regenerate everything, and verify that it works as expected. If it would be useful, I could also supply you with a test program.

Thank you for a very helpful and well-built library, and for any advice you can provide!

**Comment by emk** [comment]
Confirmed! If I remove the `upload` support from `jobs.insert`, then `configuration.query` works nicely! So the general strategy of offering both an `upload` method and a normal public `doit` method on `JobInsertCall` appears to be perfectly sound.

**Comment by emk** [comment]
Here's a version of `google-bigquery2` which provides `doit` instead of `upload`: https://github.com/faradayio/google-apis-rs/commit/7c314e54e061d2f1c42f282388b48c6310abf44a

As mentioned above, I've verified that `configuration.query` works as expected. So the proposed fix of generating both an `upload` and regular `doit` method version should work nicely.

**Comment by Byron** [comment]
Thanks so much for taking the time to write things down so that it's perfectly possible to follow you, and of course for the praise :).

I also think that the issue you encountered must be taken into consideration for the next iteration of this API generator here: #189 .

Your fix could be merged, but is likely to be overwritten next time someone runs `make update-json`. As of know, there is no mechanism for patching JSONS automatically, and possibly the generator should create 'working' versions from the originals anyway.
If you ever get an intuition how the current implementation could be fixed, please let me know or submit a PR - I am happy to merge and re-release.

**Comment by emk** [comment]
Oh, yeah, my "fix" isn't mergeable as is, because it breaks `configuration.load` while fixing `configuration.query`, `configuration.extract` and `configuration.copy`. That's just our internal workaround so I can test some other stuff.

The right fix, I think, it to change `mbuild.mako` to always generate a `pub fn doit(&self)`, and to _additionally_ generate a completely separate `upload` function when the JSON file indicates that is possible. This could do done with some copying and pasting in `mkbuild.mako` to (1) always make `doit` public with no stream or MIME argument, and to (2) fix `upload` to call a new `upload_helper` function instead of `doit`.

Does this make sense?

**Comment by Byron** [comment]
Yes it does, and the suggested fix should be relatively easy to do, too.
These days I am rather afraid to touch the 'mako' based build system,
merely because I haven't done so in a while and I know there are
essentially no tests :D.

**Comment by rossmeissl** [comment]
@Byron just curious if you plan to move forward on this?

**Comment by rossmeissl** [comment]
@Byron checking in on this

**Comment by Byron** [comment]
Apologies for the late reply.
I have currently no plan of working on this project, but welcome any contributions.

**Comment by Byron** [comment]
I am closing this issue as since then a lot has changed. All crates are now `async`, available as v2.0 on crates.io and build against the latest API spec.
Please let me know in the comments if the issue is still relevant though.
